docs/man/man1/ansible-doc.1
'\" t
.\"     Title: ansible-doc
.\"    Author: :doctype:manpage
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 12/09/2014
.\"    Manual: System administration commands
.\"    Source: Ansible 1.9
.\"  Language: English
.\"
.TH "ANSIBLE\-DOC" "1" "12/09/2014" "Ansible 1\&.9" "System administration commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ansible-doc \- show documentation on Ansible modules
.SH "SYNOPSIS"
.sp
ansible\-doc [\-M module_path] [\-l] [\-s] [module\&...]
.SH "DESCRIPTION"
.sp
\fBansible\-doc\fR displays information on modules installed in Ansible libraries\&. It displays a terse listing of modules and their short descriptions, provides a printout of their DOCUMENTATION strings, and it can create a short "snippet" which can be pasted into a playbook\&.
.SH "OPTIONS"
.PP
\fB\-M\fR \fIdirectory\fR, \fB\-\-module\-path=\fR\fIdirectory\fR
.RS 4
Add an additional directory to the default path for finding module libraries\&.
.RE
.PP
\fB\-s\fR, \fB\-\-snippet=\fR
.RS 4
Produce a snippet which can be copied into a playbook for modification, like a kind of task template\&.
.RE
.PP
\fB\-l\fR, \fB\-\-list=\fR
.RS 4
Produce a terse listing of modules and a short description of each\&.
.RE
.SH "AUTHOR"
.sp
ansible\-doc was originally written by Jan\-Piet Mens\&. See the AUTHORS file for a complete list of contributors\&.
.SH "COPYRIGHT"
.sp
Copyright \(co 2012, Jan\-Piet Mens
.sp
Ansible is released under the terms of the GPLv3 License\&.
.SH "SEE ALSO"
.sp
\fBansible\-playbook\fR(1), \fBansible\fR(1), \fBansible\-pull\fR(1)
.sp
Extensive documentation is available in the documentation site: http://docs\&.ansible\&.com\&. IRC and mailing list info can be found in file CONTRIBUTING\&.md, available in: https://github\&.com/ansible/ansible


CHANGELOG.md
Ansible Changes By Release
==========================

## 2.0 "TBD" - ACTIVE DEVELOPMENT

Major Changes:
 * Introducing the new block/rescue/always directives, allow for making task blocks and introducing exception like semantics
 * New stratergy plugins, allow to control the flow of execution of tasks per play, the default will be the same as before
 * Improved error handling, now you get much more detailed parser messages. General exception handling and display has been revamped.
 * Task includes now get evaluated during execution, end behaviour will be the same but it now allows for more dynamic includes and options.
 * First feature of the more dynamic includes is that with_ loops are now usable with them.
 * callback, connection and lookup plugin APIs have changed, some will require modification to work with new version
 * callbacks are now shipped in the active directory and don't need to be copied, just whitelisted in ansible.cfg
 * Many API changes, this will break those currently using it directly, but the new API is much easier to use and test
 * Settings are now more inheritable, what you set at play, block or role will be automatically inhertited by the contained,
   this allows for new feautures to automatically be settable at all levels, previouslly we had to manually code this
 * Many more tests, new API makes things more testable and we took advantage of it
 * big_ip modules now support turning off ssl certificate validation (use only for self signed)
 * template code now retains types for bools and Numbers instead of turning them into strings
   If you need the old behaviour, quote the value and it will get passed around as a string

Deprecated Modules (new ones in parens):
  * ec2_ami_search (ec2_ami_find)
  * quantum_network (os_network)
  * glance_image
  * nova_compute   (os_server)
  * quantum_floating_ip (os_floating_ip)

New Modules:
  * amazon: ec2_ami_copy
  * amazon: ec2_ami_find
  * amazon: ec2_eni
  * amazon: ec2_eni_facts
  * amazon: ec2_vpc_net
  * amazon: ec2_win_password
  * amazon: elasticache_subnet_group
  * amazon: iam
  * amazon: iam_policy
  * amazon: route53_zone
  * amazon: sts_assume_role
  * bundler
  * circonus_annotation
  * consul
  * consul_acl
  * consul_kv
  * consul_session
  * cloudtrail
  * cloudstack: cs_account
  * cloudstack: cs_affinitygroup
  * cloudstack: cs_facts
  * cloudstack: cs_firewall
  * cloudstack: cs_iso
  * cloudstack: cs_instance
  * cloudstack: cs_instancegroup
  * cloudstack: cs_network
  * cloudstack: cs_portforward
  * cloudstack: cs_project
  * cloudstack: cs_sshkeypair
  * cloudstack: cs_securitygroup
  * cloudstack: cs_securitygroup_rule
  * cloudstack: cs_template
  * cloudstack: cs_vmsnapshot
  * datadog_monitor
  * dpkg_selections
  * expect
  * find
  * hall
  * maven_artifact
  * openstack: os_ironic
  * openstack: os_ironic_node
  * openstack: os_client_config
  * openstack: os_floating_ip
  * openstack: os_image
  * openstack: os_network
  * openstack: os_object
  * openstack: os_security_group
  * openstack: os_security_group_rule
  * openstack: os_server
  * openstack: os_server_actions
  * openstack: os_server_facts
  * openstack: os_server_volume
  * openstack: os_subnet
  * openstack: os_volume
  * osx_defaults
  * pear
  * proxmox
  * proxmox_template 
  * puppet 
  * pushover
  * pushbullet
  * rax: rax_mon_alarm
  * rax: rax_mon_check
  * rax: rax_mon_entity
  * rax: rax_mon_notification
  * rax: rax_mon_notification_plan
  * rabbitmq_binding
  * rabbitmq_exchange
  * rabbitmq_queue
  * sensu_check
  * sensu_subscription
  * vertica_configuration
  * vertica_facts
  * vertica_role
  * vertica_schema
  * vertica_user
  * vmware_datacenter
  * vsphere_copy
  * webfaction_app
  * webfaction_db
  * webfaction_domain
  * webfaction_mailbox
  * webfaction_site
  * win_environment
  * win_scheduled_task
  * win_iis_virtualdirectory
  * win_iis_webapplication
  * win_iis_webapppool
  * win_iis_webbinding
  * win_iis_website
  * win_regedit
  * zabbix_host
  * zabbix_hostmacro
  * zabbix_screen

New Inventory scripts:
  * cloudstack
  * fleetctl
  * serf

Other Notable Changes:

## 1.9.2 "Dancing In the Street" - Jun 26, 2015

* Security fixes to check that hostnames match certificates with https urls (CVE-2015-3908)
  - get_url and uri modules
  - url and etcd lookup plugins
* Security fixes to the zone (Solaris containers), jail (bsd containers),
  and chroot connection plugins.  These plugins can be used to connect to
  their respective container types in leiu of the standard ssh connection.
  Prior to this fix being applied these connection plugins didn't properly
  handle symlinks within the containers which could lead to files intended to
  be written to or read from the container being written to or read from the
  host system instead. (CVE pending)
* Fixed a bug in the service module where init scripts were being incorrectly used instead of upstart/systemd.
* Fixed a bug where sudo/su settings were not inherited from ansible.cfg correctly.
* Fixed a bug in the rds module where a traceback may occur due to an unbound variable.
* Fixed a bug where certain remote file systems where the SELinux context was not being properly set.
* Re-enabled several windows modules which had been partially merged (via action plugins):
  - win_copy.ps1
  - win_copy.py
  - win_file.ps1
  - win_file.py
  - win_template.py
* Fix bug using with_sequence and a count that is zero.  Also allows counting backwards isntead of forwards
* Fix get_url module bug preventing use of custom ports with https urls
* Fix bug disabling repositories in the yum module.
* Fix giving yum module a url to install a package from on RHEL/CENTOS5
* Fix bug in dnf module preventing it from working when yum-utils was not already installed

## 1.9.1 "Dancing In the Street" - Apr 27, 2015

* Fixed a bug related to Kerberos auth when using winrm with a domain account.
* Fixing several bugs in the s3 module.

hacking/ansible-doc.py
#!/usr/bin/env python
# (c) 2012, Jan-Piet Mens <jpmens () gmail.com>
#
# This file is part of Ansible
#
# Ansible is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Ansible is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
#

import os
import sys
import yaml
import codecs
import json
import ast
from jinja2 import Environment, FunctionLoader, Template
import re
import optparse
import time
import datetime
from ansible import utils
from ansible import errors
from ansible.utils import module_docs
import ansible.constants as C


# Get parent directory of the directory this script lives in
MODULEDIR=os.path.abspath(os.path.join(
    os.path.dirname(os.path.realpath(__file__)), os.pardir, 'library'
    ))

T_LISTING = '''{{ "%-20s" | format(module) }}  {{ short_description | jpfunc | truncate(70) }}'''

T_SNIPPET = '''
- name: {{ short_description | jpfunc }}
  action: {{ module }}
{% for k in option_keys %}
{% set v = options[k] %}
        {{ "%-20s" | format(k + '=') }}    # {{ v.description[0] | truncate(50) | jpfunc }}
{% endfor -%}
'''

T_MODULEDOC = '''
> {{ module | upper }}  ({{filename}})

{% for desc in description %}
{{ desc | jpfunc | wordwrap(width=70) | indent(indentfirst=True) }}
{% endfor %}

{% if options %}
   Options:
{% for k in option_keys %}
{% set v = options[k] %}
   o {{ k }}
{% for desc in v.description %}{{ desc|jpfunc| wordwrap(width=60)|indent(10,indentfirst=True) }}
{% endfor %}
         {% if v.get('choices') %} Choices : {% for choice in v.get('choices',[]) %}{{ choice }}{% if not loop.last %},{%else%}.{%endif%}{% endfor %}
{% endif %}

{% endfor %}
{% endif %}

{# ." ------ NOTES #}
{% if notes %}
Note: {% for note in notes %}
{{ note | jpfunc |wordwrap(width=80) | indent(6) }}
{% endfor %}
{% endif %}

{# ------ EXAMPLES #}
{% if examples is defined %}
{% for e in examples %}
    {{ e['code'] }}
{% endfor %}
{% endif %}
'''

# There is a better way of doing this!
# TODO: somebody add U(text, http://foo.bar/) as described by Tim in #991

_ITALIC = re.compile(r"I\(([^)]+)\)")
_BOLD   = re.compile(r"B\(([^)]+)\)")
_MODULE = re.compile(r"M\(([^)]+)\)")
_URL    = re.compile(r"U\(([^)]+)\)")
_CONST  = re.compile(r"C\(([^)]+)\)")

def tty_ify(text):

    t = _ITALIC.sub("_" + r"\1" + "_", text)
    t = _BOLD.sub("*" + r"\1" + "*", t)
    t = _MODULE.sub("[" + r"\1" + "]", t)
    t = _URL.sub(r"\1", t)
    t = _CONST.sub("`" + r"\1" + "'", t)

    return t

def main():

    p = optparse.OptionParser(
        version='%prog 1.0',
        usage='usage: %prog [options] ',
        description='Show Ansible module documentation',
    )

    p.add_option("-M", "--module-dir",
            action="store",
            dest="module_dir",
            default=MODULEDIR,
            help="Ansible modules/ directory")
    p.add_option("-l", "--list",
            action="store_true",
            default=False,
            dest='list_dir',
            help='List available modules')
    p.add_option("-s", "--snippet",
            action="store_true",
            default=False,
            dest='show_snippet',
            help='Show playbook snippet for module')
    p.add_option("-m", "--module",
            action='append',
            default=[],
            dest='module_list',
            help="Add modules to process in module_dir")
    p.add_option('-V', action='version', help='Show version number and exit')

    (options, args) = p.parse_args()

    env = Environment(
        trim_blocks=True,
    )
    env.filters['jpfunc'] = tty_ify

    if options.module_dir is not None:
        utils.plugins.vars_loader.add_directory(options.module_dir)

    if options.list_dir:
        # list all modules
        ls_templ = env.from_string(T_LISTING)
        paths = utils.plugins.module_finder._get_paths()
        for path in paths:
            # os.system("ls -C %s" % (path))
            for module in os.listdir(path):
                filename = utils.plugins.module_finder.find_plugin(module)

                try:
                    doc = utils.module_docs.get_docstring(filename)
                    text = ls_templ.render(doc)
                    print text.rstrip()
                except:
                    pass
            
        sys.exit()

    tmpl = env.from_string(T_SNIPPET)
    full_tmpl = env.from_string(T_MODULEDOC)

    for module_name in options.module_list:

        filename = utils.plugins.module_finder.find_plugin(module_name)
        if filename is None:
            sys.exit("module %s not found in %s" % (module_name,
                    utils.plugins.module_finder.print_paths()))

        if filename.endswith(".swp"):
            continue

        doc = utils.module_docs.get_docstring(filename)

        if doc is None and module not in ansible.utils.module_docs.BLACKLIST_MODULES:
            sys.stderr.write("*** ERROR: MODULE MISSING DOCUMENTATION: %s ***\n" % module)
            sys.exit(1)

        if not doc is None:

            all_keys = []
            for (k,v) in doc['options'].iteritems():
                all_keys.append(k)
            all_keys = sorted(all_keys)
            doc['option_keys'] = all_keys 

            doc['filename']         = filename
            doc['docuri']           = doc['module'].replace('_', '-')
            doc['now_date']         = datetime.date.today().strftime('%Y-%m-%d')

        if options.show_snippet:
            text = tmpl.render(doc)
            print text.rstrip()
        else:
            text = full_tmpl.render(doc)
            print text.rstrip()

if __name__ == '__main__':
    main()
